pilot/pkg/serviceregistry/kube/controller/pod.go

```diff
@@ -82,46 +90,75 @@ func (pc *PodCache) onEvent(curr interface{}, ev model.Event) error {
 
 	ip := pod.Status.PodIP
 
+	// trigger the eds update when the pods are ready, or become unready
+	edsUpdateRequired := false
+
 	// PodIP will be empty when pod is just created, but before the IP is assigned
 	// via UpdateStatus.
 	if len(ip) > 0 {
 		key := kube.KeyFunc(pod.Name, pod.Namespace)
 		switch ev {
 		case model.EventAdd:
 			switch pod.Status.Phase {
-			case v1.PodPending, v1.PodRunning:
+			case v1.PodPending:
+				if key != pc.podsByIP[ip] {
+					// add to cache if the pod is pending
+					// This is for keeping the old logic
+					pc.update(ip, key)
+				}
+			case v1.PodRunning:
+				if isPodReady(pod) {
+					edsUpdateRequired = true
+				}
 				if key != pc.podsByIP[ip] {
-					// add to cache if the pod is running or pending
 					pc.update(ip, key)
 				}
 			}
 		case model.EventUpdate:
 			if pod.DeletionTimestamp != nil {
+				edsUpdateRequired = true
+				ev = model.EventDelete
 				// delete only if this pod was in the cache
 				if pc.podsByIP[ip] == key {
 					pc.deleteIP(ip)
 				}
-				ev = model.EventDelete
 			} else {
 				switch pod.Status.Phase {
-				case v1.PodPending, v1.PodRunning:
+				case v1.PodPending:
 					if key != pc.podsByIP[ip] {
-						// add to cache if the pod is running or pending
 						pc.update(ip, key)
 					}
-
+				case v1.PodRunning:
+					if isPodReady(pod) {
+						edsUpdateRequired = true
+						if key != pc.podsByIP[ip] {
+							pc.update(ip, key)
+						}
+					}
+					// When running pods become unready, we also need to trigger eds update
+					if key == pc.podsByIP[ip] && !isPodReady(pod) {
+						pc.deleteIP(ip)
+						edsUpdateRequired = true
+						ev = model.EventDelete
+					}
 				default:
 					// delete if the pod switched to other states and is in the cache
 					if pc.podsByIP[ip] == key {
 						pc.deleteIP(ip)
 					}
+					edsUpdateRequired = true
+					ev = model.EventDelete
 				}
 			}
 		case model.EventDelete:
 			// delete only if this pod was in the cache
 			if pc.podsByIP[ip] == key {
 				pc.deleteIP(ip)
 			}
+			edsUpdateRequired = true
+		}
+		if !edsUpdateRequired {
```

Comment:
I think this is not able to prevent eds update, a following ready pod can also trigger all